### PR TITLE
add doorkeeper 5.2.x compatibility

### LINKED ISSUE #18: Doorkeeper 5.1: undefined method `plaintext_token' for #<Doorkeeper::AccessGrant>
Doorkeeper `v5.1` probably has new fields that current sequel gem hasn't yet. Below error given to me on `5.1` branch of `doorkeeper` gem.

I had to downgrade to `5.0`.

**Error:**

```
undefined method `plaintext_token' for #<Doorkeeper::AccessGrant:0x00007fe4078a4ec0>
```

What steps with sequel to follow? (Or do you plan to support this gem for 5.1?)

/cc @nbulaj

### PR DESCRIPTION
**Pull request by gencer**
This fixes #18.

### CODE REVIEW

**Comment by gencer** [comment]
Only issue remains here:

```
uninitialized constant Doorkeeper::SecretStoring
```

**Comment by gencer** [comment]
On my virtual machine everything works fine but here it fails, clearly im doing something wrong. I do not want to mess anymore so leave it as is. I will use my own fork until I found a proper fix and push here. Until that, this may stay open so i can update later.

**Comment by gencer** [comment]
Also, we should drop Rails 4 and Ruby 2.1, 2.2 and 2.3.

**Comment by nbulaj** [comment]
Hi @gencer . I updated master branch to current upstream and removed old rubies / Rails, so you can rebase your PR and check if everything is OK/

**Comment by gencer** [comment]
I see `redirect_uri` _should be absolute_ and _can't be blank_ issues. But im not sure why this is happens. (`secret_strategy` fixed)

**Comment by nbulaj** [comment]
There are changes in models I believe, like allow null redirect URI and so on, I think it must be fixed too. Need to check spec/support dir and helpers, Sequel models

**Comment by gencer** [comment]
There are 86 errors and most of them are unique. I will try to fix them but oh boy so many of them :)

This is me at the moment:

tenor

**Comment by gencer** [comment]
@nbulaj, I leverage errors from 86 to 42. Mostly added missing Sequel Model Method and Variables. The rest is mostly about Doorkeeper internals and im not sure how to properly fix them. You know Doorkeeper internals better than me. I don't want to do something wrong and break.

I will be very happy if you can give a hand here. Almost all Sequel plugins works on my end with Rails-6 + latest versions of gems and only doorkeeper left unresolved :(

**Comment by nbulaj** [comment]
Need to check it out, @gencer , just don't have much time, sorry :(

**Comment by nbulaj** [comment]
just see some simple issues like:

1) `expect( AccessToken ).to receive(:find_by).with(token: hashed_token).and_return(token)` - I think out Sequel models doesn't receive `find_by` at all, we use other finder methods

2) private method 'secret_strategy' called for #<Doorkeeper::Application:0x0000000009774b98> - I think you know what we have here :)

I think there are more than one spec with such matcher that differs with Sequel model methods, so maybe try to fix the specs also. Maybe it would be better to have more ORM generic matchers like it's already done for database validation checks - https://github.com/doorkeeper-gem/doorkeeper/blob/master/spec/support/helpers/model_helper.rb#L56

**Comment by gencer** [comment]
@nbulaj, should we focus on this (`5.1`) or `doorkeeper-5.2`? Any help is appreciated. I think its time to update this gem because `5.2` released and this falls behind it. :(

**Comment by gencer** [comment]
I've fixed `MassAssignment` and token generation error. But I'm not so much familiar with Doorkeeper internals. Maybe you can have a look again @nbulaj. (I'm still trying to solve things on my end).

**Comment by nbulaj** [comment]
Hi @gencer. Sorry for the delay, I'm too busy right now. 

I'm for adding support for doorkeeper-5.2. I'll try to take a look at it when I have more free time :(

**Comment by gencer** [comment]
@nbulaj, no worries. I decreased errors from **49** to **21** now. I will be happy if you take a look at this "on your free time" -of course-.

Especially `allow_token_introspection` errors makes me unhappy atm 🤕

**Comment by nbulaj** [comment]
I've updated mainstream to latest submodule, you can use it.

Also notice that some original Doorkeeper specs can fail just because they uses specific AR features (AR querying methods) or something else that **isn't ORM agnostic**. In such case we need to fix it in Doorkeeper repo, not here.

I'll return to it when I will be free :(

**Comment by gencer** [comment]
Thanks @nbulaj, I have updated and added missing methods and decreased errors to **5**. Only **5 errors** left to get fixed.

You can have a look at your spare/free time on this. Let me know why they are happening and i will try to fix them. At the moment i have no idea why plain text token returns incorrect record even if its updated.

As you said before, it might be an issue on Doorkeeper-side. So, its on you now. Otherwise, I will fix it.

File: Gemfile [hunk 1]
```diff
@@ -13,7 +13,7 @@ end
 gem 'sequel', '>= 4.0'
 gem 'sequel_polymorphic'
 
-gem 'doorkeeper', '>= 4.0.0'
+gem 'doorkeeper', '>= 4.0.0', '< 5.2'
```
Comment (nbulaj) [inline Gemfile]: I thought we wanna to support Doorkeeper 5.2 :thinking:

**Comment by nbulaj** [comment]
I checked the error from access token model spec and see that `access_token` and `clazz.by_refresh_token(plain_refresh_token)` are different in `refresh_token` value:

```ruby
{:id=>1, :application_id=>1, :resource_owner_id=>1, :token=>"507cce35962543cbfb847db64d85b0b1ad13fa051d08c3ec963c3a4349b2d543", :refresh_token=>"e7c0cbac812042840472fd9296ee91a9f6b0682d7755c4805fabaf1e554133a0", :previous_refresh_token=>"", :expires_in=>7200, :revoked_at=>nil, :created_at=>2019-09-26 07:59:51 +0300, :scopes=>nil}

{:id=>1, :application_id=>1, :resource_owner_id=>1, :token=>"507cce35962543cbfb847db64d85b0b1ad13fa051d08c3ec963c3a4349b2d543", :refresh_token=>"plain refresh token", :previous_refresh_token=>"", :expires_in=>7200, :revoked_at=>nil, :created_at=>2019-09-26 07:59:51 +0300, :scopes=>nil}
```

Seems like it's some issue to side-effects or equality comparator (AR / Sequel)

**UPD:** I checked original AR implementation and see that it's results is the same:

```ruby
{"refresh_token"=>"e7c0cbac812042840472fd9296ee91a9f6b0682d7755c4805fabaf1e554133a0", "id"=>1, "token"=>"b426c34c9f526c4d9cbee434c4ab96938e56c53df14681fdbd961c9255900b11", "resource_owner_id"=>1, "application_id"=>1, "expires_in"=>7200, "revoked_at"=>nil, "created_at"=>Thu, 26 Sep 2019 08:11:11 UTC +00:00, "scopes"=>nil, "previous_refresh_token"=>""}

{"id"=>1, "resource_owner_id"=>1, "application_id"=>1, "token"=>"b426c34c9f526c4d9cbee434c4ab96938e56c53df14681fdbd961c9255900b11", "refresh_token"=>"plain refresh token", "expires_in"=>7200, "revoked_at"=>nil, "created_at"=>Thu, 26 Sep 2019 08:11:11 UTC +00:00, "scopes"=>nil, "previous_refresh_token"=>""}
```

**Comment by nbulaj** [comment]
I made some changes in Doorkeeper `master` so it now performs ORM-agnostic check. Update the submodule and re-initialize the specs (check Rakefile) to have proper specs. Other issues need to be checked deeper

**Comment by nbulaj** [comment]
image

strange one :thinking:

**Comment by gencer** [comment]
yeah, i saw that too. I couldn't understand why `respond_to?` receives different field.

BTW, i think i messed up my branch hehe :). There is conflict.

**Comment by gencer** [comment]
@nbulaj can you bump submodule on `master` here? I will squash commits later.

**Comment by nbulaj** [comment]
@gencer done

**Comment by gencer** [comment]
Based on `master` branch, **3** issues left.

**Comment by nbulaj** [comment]
Hey @gencer , I've fixed two more using Doorkeeper master (they exactly what I was talking about - a bad test suite and not ORM agnostic examples). You can rebase your branch

One issue least, but I can't figure out quickly why it happens

**Comment by gencer** [comment]
@nbulaj, just **one issue left**. I've squashed commits

**Comment by nbulaj** [comment]
OK, I found that the problem either in `plaintext_secret` or `to_json`, because rendered JSON for failing spec returns `nil` value for `secret`. I tried to explore the issue, but didn't succeed yet. Maybe it's related to `@raw_secret` reset or wrong secret restore using `secret_strategy`

This is the result of `Application#to_json`:

```
{:id=>1, :name=>"Example", :uid=>"yTBMkg0kUic9st7RgiPl15pAS4icK27789vNJvbzmNs", :scopes=>"", :redirect_uri=>"https://example.com", :confidential=>true, :created_at=>2019-09-27 11:42:14 +0300, :updated_at=>nil, :owner_id=>nil, :owner_type=>nil}
```

As you can see `secret` here is missed. And when we call `plaintext_secret` it returns `nil`

File: lib/doorkeeper/orm/sequel/application.rb [hunk 2]
```diff
@@ -2,14 +2,29 @@ module Doorkeeper
...
-    one_to_many :authorized_tokens, class: 'Doorkeeper::AccessToken', conditions: { revoked_at: nil }
+    one_to_many :authorized_tokens, class: "Doorkeeper::AccessToken", conditions: { revoked_at: nil }
     many_to_many :authorized_applications, join_table: :oauth_access_tokens,
                                            class: self, left_key: :id, right_key: :application_id
 
     def redirect_uri=(uris)
       super(uris.is_a?(Array) ? uris.join("\n") : uris)
     end
 
+    def plaintext_secret
```
Comment (nbulaj) [inline lib/doorkeeper/orm/sequel/application.rb:13]: It appears twice for Application class, one more in `application_mixin.rb`

Comment (gencer) [reply lib/doorkeeper/orm/sequel/application.rb:13]: Removed, will push now.

**Comment by nbulaj** [comment]
OK, @gencer , I've found an issue. When we do `to_hash` and delete `secret`, Sequel model reset it's value for the instance and then we have an empty value. So you need to do a `#dup` first:

```ruby
    def to_json(options = nil)
      hash = to_hash.dup
      hash.delete(:secret)
      hash.merge(secret: plaintext_secret)
        .to_json(options)
    end
```

Also see, let's rename `elems` to `hash`, I personally don't like such abbreviations

**Comment by gencer** [comment]
> OK, @gencer , I've found an issue. When we do `to_hash` and delete `secret`, Sequel model reset it's value for the instance and then we have an empty value. So you need to do a `#dup` first

```ruby
    def to_json(options = nil)
      elems = dup.to_hash
      elems.delete(:secret)
      elems.merge(secret: plaintext_secret)
        .to_json(options)
    end
```

Are we talking about this one?

**Comment by nbulaj** [comment]
Already updated. Yep, exactly! See my example above :point_up:

**Comment by gencer** [comment]
pushed now. let's watch the Travis.

**Comment by gencer** [comment]
Success 👍 🎉🎉🎉🎉🎉

**Comment by nbulaj** [comment]
I think we can merge it now :+1: 

Thanks @gencer for such a great work!

**Comment by gencer** [comment]
Thank you so much @nbulaj for your help and patient. ❤

Have a wonderful weekend!

**Comment by gencer** [comment]
@nbulaj just a question.

Can we expect a new version soon? Currently im on `master` branch which is non-semantic.

**Comment by nbulaj** [comment]
Give me a few days please, maybe tomorrow or evrn today :)

**Comment by gencer** [comment]
Ahh. Just asked. No worries. 😜  Have it your way.

**Comment by nbulaj** [comment]
Actually I found time now :)

> Successfully registered gem: doorkeeper-sequel (2.1.0)

I think we'll face more issues, create issues/PR and we'll go with them :handshake:

**Comment by gencer** [comment]
Thank you!

I'm always here. I will do my best! 🤝

**Pull request merged**